src/dmd/dmodule.d

```diff
@@ -405,39 +405,51 @@ extern (C++) final class Module : Package
     size_t nameoffset;          // offset of module name from start of ModuleInfo
     size_t namelen;             // length of module name in characters
 
-    extern (D) this(const ref Loc loc, const(char)* filename, Identifier ident, int doDocComment, int doHdrGen)
+    extern (D) this(const ref Loc loc, const(char)[] filename, Identifier ident, int doDocComment, int doHdrGen)
     {
         super(loc, ident);
-        const(char)* srcfilename;
+        const(char)[] srcfilename;
         //printf("Module::Module(filename = '%s', ident = '%s')\n", filename, ident.toChars());
         this.arg = filename;
-        srcfilename = FileName.defaultExt(filename, global.mars_ext);
-        if (global.run_noext && global.params.run && !FileName.ext(filename) && FileName.exists(srcfilename) == 0 && FileName.exists(filename) == 1)
+        srcfilename = FileName.defaultExt(filename, global.mars_ext.toDString());
+        if (global.run_noext && global.params.run &&
+            !FileName.ext(filename) &&
+            FileName.exists(srcfilename) == 0 &&
+            FileName.exists(filename) == 1)
         {
-            FileName.free(srcfilename);
+            FileName.free(srcfilename.ptr);
             srcfilename = FileName.removeExt(filename); // just does a mem.strdup(filename)
         }
-        else if (!FileName.equalsExt(srcfilename, global.mars_ext) && !FileName.equalsExt(srcfilename, global.hdr_ext) && !FileName.equalsExt(srcfilename, "dd"))
+        else if (!FileName.equalsExt(srcfilename, global.mars_ext.toDString()) &&
+                 !FileName.equalsExt(srcfilename, global.hdr_ext.toDString()) &&
+                 !FileName.equalsExt(srcfilename, "dd"))
         {
-            error("source file name '%s' must have .%s extension", srcfilename, global.mars_ext);
+            error("source file name '%s' must have .%s extension", srcfilename.ptr, global.mars_ext);
```

Comment:
This relies on .ptr being 0 terminated. But I don't see where that guarantee is made.

Reply:
These strings come from FileName and its methods, which make that guarantee.  I'll continue to try to switch stuff to slices so we can eliminate spamming .ptr everywhere.

Reply:
Please pass the length and pointer separately to `printf`-like functions. Then 0 termination doesn’t matter.